### ISSUE TITLE
[BlenderBot 1.0] Finetuning BlenderBot 1.0 on PersonaChat dataset

### ISSUE THREAD

**Issue by passing2961** [facebookresearch/ParlAI#4581]
Hi,

Currently, I try to fine-tune blenderbot 1.0 on personachat dataset (i.e., `self_original` version) with default hyperparameter setting provided by ParlAI, which is the purpose of replicating the results reported in the original paper (i.e., Recipes for building an open-domain chatbot). In this paper, the authors said that "Models trained on this task are thus conditioned on the persona and the dialogue history, which are concatenated". So, what this means is that for every training instances, the input to the encoder is the concatenated version of persona and dialogue history? But when I fine-tuned, I printed out actual inputs in `history` class (see below example).

```
{'text': 'your persona: i am now looking for a new job.\nyour persona: my favorite movie is friday.\nyour persona: i am a doctor.\nyour persona: my daughter is a child prodigy.\nhello , i am sitting here with my dog . how are you ?', 'reward': 0, 'label_candidates': ['what do you do for a living ? mechanical engineering is my day job .', 'do you like those colors together ? do they go good ?', 'hi , i am am very good', 'i am a fisherman do you like your job', 'i am fine thank you , how about you ?', 'well i do not know if it counts i only work at a book shop', 'oh god why is there fire , why is there fire ? this is not nascar', 'nice to hear . i am doing great . thanks for asking', 'ok , just making sure . i am 40 .', 'wow , that is awesome what year of law school are you in ?', 'yes , so italian food is big in our family !', 'oh that is very rewarding', 'where do you want to go , how much will you paid me to drive you ?', 'i will admit , it is not a common thing , but it is a great deal of fun .', 'hi , how are y o u ?', 'i am 19 saving to move out of my parents !', 'i am alright feeling old i am 45', 'hi i love to dance , i take lessons every week', 'that is good ! what kind of business are you in ? you hiring ?', 'i am well friend . looking for new employment at the moment .'], 'episode_done': False, 'id': 'personachat:self_original', 'eval_labels': ['i am well friend . looking for new employment at the moment .']}
```
```
{'text': 'i just got done watching a horror movie', 'reward': 0, 'label_candidates': ['why have you not sent help ? ! the scorpions are stinging my legs ! ree ! ! ! ! ! ! !', 'that is great i am expecting twins in two months . will these be your first kids ?', 'do you live on a farm or ranch ?', 'hi how are you doing tonight i am fine .', 'i would love to see her do that .', 'i do not . but i am so glad you do something that brings you joy', 'it is hard , buy my dog keeps me company . do you have dogs ?', 'sounds like a good plan , what would you like to teach ?', 'i like rap music and i also produce for music artists .', 'where do you work ?', 'i broke my arm so i can not drink coffee', 'i meant mickey . cool i have a lot of friends and love the playground . do you ?', 'oh , yes . i wish i could skate to school . i ride the bus instead .', 'that is my idea of heaven . i love bunnies ! i donate to a bunny charity too .', 'i work on a freelance basis as an author , blogger and affiliate marketer . and you ?', 'like me ? would you go on vacation to the beach with me', 'i do not let myself watch tv', 'we can go for a trade that sounds awesome', 'what part of the world is bratislava ?', 'i rather read , i have read about 20 books this year .'], 'episode_done': False, 'id': 'personachat:self_original', 'eval_labels': ['i rather read , i have read about 20 books this year .']}
```
In the first example, the input to the encoder is the concatenated version. But, in the second example, there is no persona information in `text` field. 

Thus, in summary, my questions are as follows:
(1) Can you tell me more detailed information about how to fine-tune blenderbot on personachat dataset?
(2) When you fine-tune the blenderbot, which version of personachat was used? (e.g., `self_original`, `both_original` or `revised` version)

Best regards,

**Comment by klshuster** [comment]
The model is provided with successive turns of a conversation; only the *new* turns are provided. Note that `episode_done` is set to `False` in each of these examples - this informs the `history` class to not clear its current output and instead append the incoming to text to the full conversational context. This entire context, then, is sent to the model as a new-line delimited string (well, a tokenized version)

If you set `-t convai2:normalized`, you will get the same data we used to train BlenderBot 1. This corresponds to the `self_original` task

**Comment by passing2961** [comment]
Thanks for kindly responding to my question. I understand the entire process and how it works.

I have the last two questions. 
(1) In BlenderBot 1, it seems you don't apply speaker role embeddings. When I saw the code, it only used the positional embeddings. Do I understand correctly?
(2) When the successive turns of a conversation are too long to use all the information as the input of the model (meaning that the truncation is needed), are there any effective truncation strategies that are specialized in the persona-based dialogue generation task? 

Best regards,

**Comment by klshuster** [comment]
1. That is correct
2. BlenderBot 1 has a 128-token truncation length; we've generally found that keeping the latest 128 tokens in context to be the most effective at maintaining appropriate conversational skills - although this does indeed truncate the personas
